### PR TITLE
Add ZIP-244 signature hash support

### LINKED ISSUE #2051: ZIP-244: Implement sighash for Non-Malleable Transaction Ids
**Is your feature request related to a problem? Please describe.**

ZIP-244 defines a new transaction digest algorithm currently planned to be applied from the NU5 network upgrade onward.

ZIP-244 creates non-malleable transaction identifiers that commit to all transaction data except for attestations to transaction validity. (That is, signatures and proofs.)

ZIP-244 also adds an updated sighash algorithm.

**Describe the solution you'd like**

We should implement or integrate this new transaction digest algorithm in Zebra as per:
* https://zips.z.cash/zip-0244
* https://zips.z.cash/zip-0225#modifications-to-zip-244

This will include:

Sighash updates
- update `sighash` based on ZIP-244 with Orchard changes using librustzcash
  - see https://github.com/zcash/librustzcash/pull/319

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

ZIP-244 and ZIP-225 specify a new signature hash that reuses many digests computed during the transaction ID digest computation. We need to implement it for network upgrade 5.

## Solution

Use librustzcash to compute it, to save implementation time.

I had to make some changes to librustzcash in order to be able to call it. I'm discussing these with the ECC team to incorporate in the library. Currently this PR makes zebra point to a branch in our fork of librustzcash with those changes.

The test is still not passing, I haven't investigated the reason yet. But it may be better to wait for ECC to run on their test vectors first.

The solution is not particularly efficient because it doesn't reuse the digest computation from the transaction ID. But we have discussed previously that we should make it work first.

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Review

It doesn't seem urgent but I think it's best to review it in this sprint.

@teor2345 is already familiar with the librustzcash integration, but @dconnolly may be interested in taking a look

One particular point I'd like feedback is the location of the test. It's on `vectors.rs` with the other ZIP-244 tests, but all previous sighashes tests are in `sighash.rs`. Should I move it there?

## Related Issues

Closes #2051 

This was implemented on top of #2050, so that should be merged first.

## Follow Up Work

Follow up with librustzcash to make sure the test passes.

#2183: also use librustzcash for V4 and earlier

### CODE REVIEW

**Comment by teor2345** [comment]
Hey @conradoplg it looks like this PR is based on #2129, so I've changed the base branch.

GitHub will automatically rebase the PR to main when #2129 merges. But you might need to rebase force-push the branch if we do a squash.

**Comment by teor2345** [comment]
> One particular point I'd like feedback is the location of the test. It's on `vectors.rs` with the other ZIP-244 tests, but all previous sighashes tests are in `sighash.rs`. Should I move it there?

We try to put tests in their own files to speed up compilation. So please move all the tests from sighash.rs to vectors.rs.

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (teor2345) [inline zebra-chain/src/primitives/zcash_primitives.rs:35]: If you want to do all the sighashes through librustzcash, that would also be ok.

Let's open a follow-up ticket for that?

File: zebra-chain/src/transaction/tests/vectors.rs
Comment (teor2345) [inline zebra-chain/src/transaction/tests/vectors.rs]: Can we test against the Zebra v4 sighash impl for all the v4 transactions in BLOCKS?

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (teor2345) [inline zebra-chain/src/primitives/zcash_primitives.rs]: I think we should just use `Hash(pub [u8; 32])` as the return type.

Otherwise, we are exposing an implementation-defined type in our API. (And the librustzcash API.)

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good, I made some suggestions about return types and testing

**Comment by conradoplg** [comment]
> > One particular point I'd like feedback is the location of the test. It's on `vectors.rs` with the other ZIP-244 tests, but all previous sighashes tests are in `sighash.rs`. Should I move it there?
> 
> We try to put tests in their own files to speed up compilation. So please move all the tests from sighash.rs to vectors.rs.

Sure! But the tests access private methods (e.g. `hash_header`) so from what I understand I'd need to make them `pub(super)` which does not seem great. What do you think?

File: zebra-chain/src/transaction/tests/vectors.rs
Comment (conradoplg) [reply zebra-chain/src/transaction/tests/vectors.rs]: I've added the test; I had to do a small refactoring. However, the test is failing... I'll investigate a bit, tomorrow

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (conradoplg) [reply zebra-chain/src/primitives/zcash_primitives.rs]: Sounds good, done!

**Comment by teor2345** [comment]
> Sure! But the tests access private methods (e.g. `hash_header`) so from what I understand I'd need to make them `pub(super)` which does not seem great. What do you think?

`pub(super)` or `pub(crate)` is fine for tests. (We try not to expose implementation details outside the crate, but we can be flexible inside the crate.)

File: zebra-chain/src/transaction/tests/vectors.rs
Comment (conradoplg) [reply zebra-chain/src/transaction/tests/vectors.rs]: Apparently the issue is in librustzcash. Changing it to what I think it's the expected behaviour, all tests pass for the older transactions. I've already raised the issue with the ECC team to confirm.

(The v5 is still failing, I haven't investigated it)

**Comment by conradoplg** [comment]
> > Sure! But the tests access private methods (e.g. `hash_header`) so from what I understand I'd need to make them `pub(super)` which does not seem great. What do you think?
> 
> `pub(super)` or `pub(crate)` is fine for tests. (We try not to expose implementation details outside the crate, but we can be flexible inside the crate.)

Makes sense! I've moved the tests.

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (conradoplg) [reply zebra-chain/src/primitives/zcash_primitives.rs:35]: Opened in #2183

**Comment by conradoplg** [comment]
Rebased with https://github.com/ZcashFoundation/zebra/pull/2129, updated test to not skip Orchard txs. V5 test still not passing.

**Comment by conradoplg** [comment]
Test fixed and passing.

Added a round-trip test suggested in #2129

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:57]: Could this be a method on `Transaction` such as `Transaction::into_librustzcash(network_upgrade: NetworkUpgrade) -> Result<zcash_primitives::transaction::Transaction, io::Error>` or something?  I'm thinking of: https://rust-lang.github.io/api-guidelines/naming.html#c-conv

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:66]: 👍

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [inline zebra-chain/src/primitives/zcash_primitives.rs:74]: 👍

File: zebra-chain/src/transaction/sighash.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/sighash.rs]: ~Is this not the same as the `blake2b_simd::Hash` being returned from https://doc-internal.zebra.zfnd.org/zebra_chain/transaction/enum.Transaction.html#method.sighash ?~
Nevermind I see that's been supplanted here

File: zebra-chain/src/transaction.rs
Comment (dconnolly) [inline zebra-chain/src/transaction.rs]: This is short and clear what kind of Hash we're talking about

```suggestion
    ) -> sighash::Hash {
```

File: zebra-chain/src/transaction/sighash.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/sighash.rs]: So, the proper name is literally [`SIGHASH`](https://zips.z.cash/protocol/protocol.pdf#sighash), and I don't think I ever hear anyone say 'signature hash' when they are talking about the result of applying the `SIGHASH` algorithm to a transaction, so we may just want to call it `SigHash` everywhere

File: zebra-chain/src/transaction/sighash.rs
Comment (dconnolly) [inline zebra-chain/src/transaction/sighash.rs]: FYI types implementing the `std::hash::Hash` trait are '...able to be hashed with an instance of Hasher', which is true here but just making sure that's clear

**Review by dconnolly — CHANGES_REQUESTED** [review]
Some naming questions and other stuff 🤩

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (conradoplg) [reply zebra-chain/src/primitives/zcash_primitives.rs:57]: My reasoning is that I wanted to keep every usage of `librustzcash/zcash_primitives` inside this file.

This particular function is just a helper for the  `TryFrom` implementation above, so if I follow your suggestion, I would remove the `TryFrom` and create a `into_librustzcash`?

Maybe a more generic question is: when should we use `TryFrom` and when should we use a `into_` method?

File: zebra-chain/src/transaction/sighash.rs
Comment (conradoplg) [reply zebra-chain/src/transaction/sighash.rs]: Yeah that does not apply here. At the time I based this on `block::hash::Hash` but forgot to review it later. Thanks for catching it! Done in 463f5f29

File: zebra-chain/src/transaction/sighash.rs
Comment (conradoplg) [reply zebra-chain/src/transaction/sighash.rs]: I was confused about the terminology because Bitcoin seems to call ["sighash" the sighash flags](https://developer.bitcoin.org/glossary.html#term-Sighash), not the hash itself (which is called "Signed hash"?). ZIP-244 also uses "signature digest".

But yeah, using yet another name is even more confusing :sweat_smile: , I've changed it to `SigHash`. Done in 463f5f29

File: zebra-chain/src/transaction.rs
Comment (conradoplg) [reply zebra-chain/src/transaction.rs]: Done in 463f5f29

File: zebra-chain/src/primitives/zcash_primitives.rs
Comment (dconnolly) [reply zebra-chain/src/primitives/zcash_primitives.rs:57]: Oh! I totally missed that. You are right, ignore me.

File: zebra-chain/src/transaction/sighash.rs
Comment (dconnolly) [reply zebra-chain/src/transaction/sighash.rs]: There's so much overloading of all these terms over the history of Bitcoin and now Zcash 🙃 📜 📚

**Review by dconnolly — APPROVED** [review]
lgtm! 🌠

**Review by dconnolly — DISMISSED** [review]

**Comment by conradoplg** [comment]
@teor2345 or @dconnolly can you reapprove? I had to solve merge conflicts after the txid PR was merged

**Review by dconnolly — APPROVED** [review]

**Comment by conradoplg** [comment]
It seems @teor2345's approval is required, but I swear the merging button was available after @dconnolly approved it the first time :thinking: (It seems I can dismiss Teor's request for changes but it doesn't feel like a good practice, so I'll wait for now)

**Review by teor2345 — APPROVED** [review]
This is fine, please don't block on me.

**Pull request merged**
